plugins/gradle/src/org/jetbrains/plugins/gradle/service/execution/GradleProgressIndicatorWrapper.kt

```diff
@@ -0,0 +1,68 @@
+// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.jetbrains.plugins.gradle.service.execution
+
+import com.intellij.openapi.externalSystem.model.task.ExternalSystemTaskId
+import com.intellij.openapi.externalSystem.model.task.ExternalSystemTaskNotificationEvent
+import com.intellij.openapi.externalSystem.model.task.ExternalSystemTaskNotificationListener
+import com.intellij.openapi.externalSystem.model.task.ExternalSystemTaskNotificationListenerAdapter
+import com.intellij.openapi.externalSystem.service.internal.ExternalSystemTaskProgressIndicatorUpdater
+import com.intellij.openapi.externalSystem.util.ExternalSystemBundle
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
+import org.jetbrains.annotations.Nls
+import org.jetbrains.plugins.gradle.service.execution.GradleProgressIndicatorEventHelper.areGradleBuildProgressEventsSupported
+import org.jetbrains.plugins.gradle.settings.GradleExecutionSettings
+import org.jetbrains.plugins.gradle.util.GradleConstants
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
+
+internal class GradleProgressIndicatorWrapper(
+  private val id: ExternalSystemTaskId,
+  private val effectiveSettings: GradleExecutionSettings,
+  private val listener: ExternalSystemTaskNotificationListener
+) {
+
+  fun runWithProgressIndicator(consumer: Consumer<ExternalSystemTaskNotificationListener>) {
+    if (!areGradleBuildProgressEventsSupported(effectiveSettings)) {
+      // Don't run with progress indicator if build progress events are not supported
+      consumer.accept(listener)
+      return
+    }
+
+    val progressIndicator = AtomicReference<ProgressIndicator>()
+    val countDownLatch = CountDownLatch(1)
+    ProgressManager.getInstance().run(object : Task.Backgroundable(id.findProject(), "Gradle build", false) {
```

Comment:
this steals a thread just to create an indicator. Please, do not do this. I think you now know better ways (as GradleDownloadProgressListenerImpl).